torch/distributed/device_mesh.py

```diff
@@ -180,50 +134,48 @@ def create_flatten_mesh(
                     f"Please specify another valid mesh_dim_name.",
                 )
 
+            flattened_mesh_layout = device_mesh._layout.coalesce()
             # Quick return if the flatten mesh has been created before.
             if (
                 root_mesh in self.root_to_flatten_mapping
                 and mesh_dim_name in self.root_to_flatten_mapping[root_mesh]
             ):
                 if (
-                    tuple(flatten_dims_in_root)
-                    == self.flatten_name_to_root_dims[root_mesh][mesh_dim_name]
+                    flattened_mesh_layout
+                    == self.flatten_name_to_root_layout[root_mesh][mesh_dim_name]
                 ):
                     return self.root_to_flatten_mapping[root_mesh][mesh_dim_name]
                 else:
                     raise RuntimeError(
                         f"Flatten mesh with mesh_dim_name {mesh_dim_name} has been created before, "
-                        f"Please specify another valid mesh_dim_name.",
+                        f"Please specify another valid mesh_dim_name."
                     )
 
-            flattened_mesh_dim_size = math.prod(device_mesh.mesh.size())
-
-            remained_dims_in_root = list(range(root_mesh.mesh.ndim))
-            for flatten_dim_in_root in flatten_dims_in_root:
-                remained_dims_in_root.remove(flatten_dim_in_root)
-
-            pg_ranks_by_dim = root_mesh.mesh.permute(
-                *remained_dims_in_root, *flatten_dims_in_root
-            ).reshape(-1, flattened_mesh_dim_size)
-
             cur_rank = root_mesh.get_rank()
+            # Due to the limitation of ProcessGroup api, we need to start from root mesh so that all ranks call the
+            # new_group api to avoid potential hang.
```

Comment:
I feel bad that we need to make the DeviceMesh code more complicated in order to deal with these low-level concerns. Moreover, if the backend supports splitting (e.g., NCCL) then all of this isn't even actually needed!

May I suggest that we add a helper function that "splits" the root PG into an arbitrary "tiling" (i.e., N subgroups which cover everything) which abstracts away these complexities by internally deciding whether to call split or do this for loop?